### PR TITLE
feat(fastify): Integrate apollo-fastify plugin #626 [WIP]

### PR DESCRIPTION
**Pull request by addityasingh**
Refer #626 , #1013 

TODO:

* [ ] Update CHANGELOG.md with your change (include reference to issue & this PR)
* [ ] Make sure all of the significant new logic is covered by tests
* [x] Rebase your changes on master so that they can be merged easily

### CODE REVIEW

File: packages/apollo-server-fastify/src/__tests__/fastifyApollo.test.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  });
+
+  it('throws an error if called with argument not equal to 3', function() {
+    expect(() => (<any>graphqlFastify)({}, { graphqlOptions: {} })).toThrow(
+      'Apollo Server expects exactly 3 argument, got 2',
+    );
+  });
+});
+
+// Uncomment this to see the breaking tests
```
Comment (addityasingh) [inline packages/apollo-server-fastify/src/__tests__/fastifyApollo.test.ts]: The tests for basic request to graphql with a `POST` request in this suite seem to be breaking with `400 ` status code with below message. 

```
 {"statusCode":400,"error":"Bad Request","message":"Request body size did not match Content-Length"}
```

File: packages/apollo-server-fastify/src/ApolloServer.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        response: ServerResponse,
+        next: any,
+      ) {
+        if (request.path !== path) {
+          return next();
+        }
+
+        if (this.playgroundOptions && request.method === 'get') {
+          // perform more expensive content-type check only if necessary
+          const accept = parseAll(request.headers);
```
Comment (mcollina) [inline packages/apollo-server-fastify/src/ApolloServer.ts]: I would use https://github.com/fastify/fastify-accepts instead.

File: packages/apollo-server-fastify/src/ApolloServer.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            };
+
+            response.setHeader('Content-Type', 'text/html');
+            return response.end(
+              renderPlaygroundPage(playgroundRenderPageOptions),
+            );
+          }
+        }
+        return next();
+      }.bind(this),
```
Comment (mcollina) [inline packages/apollo-server-fastify/src/ApolloServer.ts]: I would recommend doing this in a `beforeHandler`  hook instead.

File: packages/apollo-server-fastify/src/__tests__/ApolloServer.test.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  let httpServer: http.Server;
+
+  async function createServer(
+    serverOptions: Config,
+    options: Partial<ServerRegistration> = {},
+  ) {
+    server = new ApolloServer(serverOptions);
+    app = fastify();
+
+    server.applyMiddleware({ ...options, app });
```
Comment (mcollina) [inline packages/apollo-server-fastify/src/__tests__/ApolloServer.test.ts]: A more idiomatic fastify approach is:

```js
app = fastify();
app.register(apollo, { ...options });
```

File: packages/apollo-server-fastify/src/fastifyApollo.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      });
+
+      reply
+        .type('application/json')
+        .code(200)
+        .header(
+          'Content-Length',
+          Buffer.byteLength(JSON.stringify(graphqlResponse), 'utf8'),
+        )
+        .send(JSON.parse(graphqlResponse));
```
Comment (mcollina) [inline packages/apollo-server-fastify/src/fastifyApollo.ts]: This is really not clear. Why are you serializing the response back-and-forth again? You are probably better off as not parsing/stringifying both the body and the request/response. This likely halves the throughput of this solution.

File: packages/apollo-server-fastify/src/__tests__/fastifyApollo.test.ts (hunk 1 above)
Comment (mcollina) [reply packages/apollo-server-fastify/src/__tests__/fastifyApollo.test.ts]: You should update to Fastify 1.12.1, this should be fixed.

File: packages/apollo-server-fastify/src/ApolloServer.ts (hunk 2 above)
Comment (addityasingh) [reply packages/apollo-server-fastify/src/ApolloServer.ts]: Done

File: packages/apollo-server-fastify/src/ApolloServer.ts (hunk 3 above)
Comment (addityasingh) [reply packages/apollo-server-fastify/src/ApolloServer.ts]: Done. Added a `preHandler`

File: packages/apollo-server-fastify/src/__tests__/ApolloServer.test.ts (hunk 4 above)
Comment (addityasingh) [reply packages/apollo-server-fastify/src/__tests__/ApolloServer.test.ts]: This has some Typescript errors due to the apollo interfaces. Will try to take a deeper look into this later

File: packages/apollo-server-fastify/src/fastifyApollo.ts (hunk 5 above)
Comment (addityasingh) [reply packages/apollo-server-fastify/src/fastifyApollo.ts]: Removing the serialization breaks most of the tests. I am assuming that the response need to be serialized.

File: packages/apollo-server-fastify/src/ApolloServer.ts (hunk 2 above)
Comment (mcollina) [reply packages/apollo-server-fastify/src/ApolloServer.ts]: I don't think this is working right now, the usage is different.

File: packages/apollo-server-fastify/src/fastifyApollo.ts (hunk 5 above)
Comment (mcollina) [reply packages/apollo-server-fastify/src/fastifyApollo.ts]: I mean that you are getting JSON via `runHttpQuery`, which I imagine it is parsed and stringified again, or maybe I'm mistaken by the API.

File: packages/apollo-server-fastify/src/ApolloServer.ts (hunk 2 above)
Comment (addityasingh) [reply packages/apollo-server-fastify/src/ApolloServer.ts]: @mcollina I need a `next()` method along with `async/await` but based on the docs we can't use `next()` along with `async/await`. What is the alternative in this case?

File: packages/apollo-server-fastify/src/fastifyApollo.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    handler: async (_: any, reply: FastifyReply<OutgoingMessage>) => {
+      reply
+        .code(405)
+        .header('allow', 'GET, POST')
+        .send();
+    },
+  });
+
+  if (next && typeof next === 'function') {
+    next();
```
Comment (jimmiehansson) [inline packages/apollo-server-fastify/src/fastifyApollo.ts]: A question to clarify? Shouldn't this return back to the previous context if the callback is invoked, if `next()` is available. 
E.g. I am assuming at this point that the invocation of the callback is what matters, which else wise could render undefined at some point?
```js
if (next && typeof next === 'function') {
  next();
  return;
}
```

File: packages/apollo-server-fastify/src/ApolloServer.ts (hunk 2 above)
Comment (mcollina) [reply packages/apollo-server-fastify/src/ApolloServer.ts]: Why do you need it?

**Comment by cemremengu** [comment]
@addityasingh any updates here? I would like to help to speed things up.

**Comment by abernix** [comment]
@addityasingh As I mentioned in #2280, I didn't realize we had two PRs in flight that built the same integration.  The names were so similar, that I just realized after I merged that PR. Furthermore, you're the actual current maintainer of the `apollo-server-fastify` package on npm and you've already gone to the effort of getting it, as you mention above.

You've put some effort into this integration, in both in v1 and v2, and I think you deserve commit credit, despite the fact that https://github.com/apollographql/apollo-server/pull/1971 ended up landing.  I merged this PR into #2280, preferring that branches implementation, but you'll see your commit credit.

I'll close this now, with it being superseded by #1971 and @2280, but thanks for doing this originally.  I'm working on catching up on the back-log to help reduce the chance of this in the future.

We're still looking for your blessing on the `apollo-server-fastify` package name, but #1971 	talks about that the most.  Thanks again!

**Comment by addityasingh** [comment]
@abernix Yeah. I opened 2 PRs (#1013 , #1760 ) to get the `fastify` integration in `apollo`, the first of which was opened on the 1st of May 2018. But none of them could be merged for so long and I couldn't keep them updated recently. But I am fine with having the PR #1971 by @rkorrelboom merged. Like I mentioned [here](https://github.com/apollographql/apollo-server/pull/2280#issuecomment-462808486), I have also transferred the ownership of the package to you and `apollo-bot`

**Pull request closed without merging**